staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,594 @@
+	// If the following calls produce an error we only log them and continue
+	// with monitoring the invoice settlement and listen for the htlc tx to
+	// be mined. Otherwise, if we returned on an error in the following, we
+	// wouldn't be able to monitor the htlc tx.
+	// Also, if the following code fails we couldn't sign the server's
+	// sweepless sweep, which wouldn't hinder the swap to succeed.
+	f.sweeplessServerNonces, err = toNonces(
+		sigPushResp.SweeplessServerNonces,
+	)
+	if err != nil {
+		log.Warnf("unable to convert server nonces: %v", err)
+	}
+	sweeplessAddress, err := btcutil.DecodeAddress(
+		sigPushResp.SweeplessSweepAddr, f.cfg.ChainParams,
+	)
+	if err != nil {
+		log.Warnf("unable to decode sweep address: %v", err)
+	}
+	f.loopIn.SweeplessSweepAddress = sweeplessAddress
+	f.loopIn.SweeplessSweepFeeRate = chainfee.SatPerKWeight(
+		sigPushResp.SweeplessFeeRate,
+	)
+
+	return OnHtlcTxSigned
+}
+
+// MonitorInvoiceAndHtlcTxAction is called after the htlc tx has been signed by
+// us. The server from here on has the ability to publish the htlc tx. If the
+// server publishes the htlc tx without paying the invoice, we have to monitor
+// for the timeout path and sweep the funds back to us. If, while waiting for
+// the htlc timeout, our invoice gets paid, the swap is considered successful,
+// and we can stop monitoring the htlc confirmation and continue to sign the
+// sweepless sweep.
+func (f *FSM) MonitorInvoiceAndHtlcTxAction(_ fsm.EventContext) fsm.EventType {
+	// Subscribe to swap invoice settlement.
+	invoiceUpdateChan, invoiceErrChan, err := f.cfg.InvoicesClient.SubscribeSingleInvoice( //nolint:lll
+		f.ctx, f.loopIn.SwapHash,
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	htlc, err := f.loopIn.getHtlc(f.cfg.ChainParams)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	// Subscribe to htlc tx confirmation.
+	htlcConfChan, htlcErrConfChan, err := f.cfg.ChainNotifier.RegisterConfirmationsNtfn( //nolint:lll
+		f.ctx, nil, htlc.PkScript, defaultConfTarget,
+		int32(f.loopIn.InitiationHeight),
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	// Subscribe to new blocks.
+	blockChan, blockChanErr, err := f.cfg.ChainNotifier.RegisterBlockEpochNtfn(f.ctx) //nolint:lll
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	htlcConfirmed := false
+
+	for {
+		select {
+		case <-htlcConfChan:
+			htlcConfirmed = true
+
+		case err = <-htlcErrConfChan:
+			return f.HandleError(err)
+
+		case currentHeight := <-blockChan:
+			cancelInvoice := func() {
+				log.Errorf("timeout waiting for invoice to be "+
+					"paid, canceling invoice: %v",
+					f.loopIn.SwapHash)
+
+				err = f.cfg.InvoicesClient.CancelInvoice(
+					f.ctx, f.loopIn.SwapHash,
+				)
+				if err != nil {
+					log.Warnf("unable to cancel invoice "+
+						"for swap hash: %v %v",
+						f.loopIn.SwapHash, err)
+				}
+			}
+
+			htlcTimedOut := f.loopIn.isHtlcTimedOut(currentHeight)
+
+			if !htlcConfirmed {
```

Comment:
nit: perhaps start with checking if timed out and cancel, then depending on confirmation status transition if needed (to reduce nesting).

Reply:
I've removed the nested clauses.